rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,265 @@
+
+```text
+emit_log(value: Object)
+```
+
+This function will cause the object passed as value to be emitted at that point and flushed downstream. The emitted log will have its metadata copied from the input event.
+
+Additionally, an `emit_root` (we can work on the naming) config option will be added to the `remap` transform to configure whether `.` is emitted after the transform runs. It will default to `true` to preserve the current behavior but can be set to `false` by users to supress this behavior. Admittedly, I'm not wild about introducing this additional config option, but I'm not seeing another great alternative.
+
+This will be able to be combined with the iteration mechanism that will be introduced [#6031](https://github.com/timberio/vector/issues/6031) to emit an unknown number of events. Naively this might look something like:
+
+```text
+for stooge in .stooges
+   emit_log(stooge)
+end
+```
+
+In the future we can also add functions for emitting metrics like:
+
+```text
+emit_counter(namespace: String, name: String, timestamp: Timestamp, value: Float, kind: "absolute"|"relative")
+```
+
+I considered having just an `emit_metric()` but it would require users to pass in objects that match exactly the internal representation we have for metrics.
+
+The `remap` tranform would gain an extra configuration option:
+
+```text
+emit_root = true/false # default false
+```
+
+When `emit_root` is `true`, the value of `.` will be emitted at the end of the remap program. When `emit_root` is false, the value of `.` will not be emitted. Instead users should use the `emit_log` function to emit.
+
+We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.
+
+### Modifying remap to accept setting the root object to an array
+
+https://github.com/timberio/vector/issues/6988
+
+This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.
+
+This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.
+
+Using a separate `explode` transform keeps the responsibilities of the transform more clear and avoids having to refactor the `remap` transform to decouple `.` from the underlying Vector `Event` object; though we may still want to do this in the future anyway.
+
+## Open Questions
+
+1. Do we want the additional `unnest` transform? I included it based on [feedback that using VRL was offputting for simple cases](https://github.com/timberio/vector/pull/7038#issuecomment-817043491), but maybe the newly proposed `remap` modifications are easy enough to use that we don't need the additional transform.
```

Comment:
Ha, the current Ben disagrees with the past Ben. My stance has changed. Assigning `.` to an array has always struck me as the simplest approach.